presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/joins/JoinGraph.java

```diff
@@ -200,15 +218,28 @@ private JoinGraph joinWith(JoinGraph other, List<JoinNode.EquiJoinClause> joinCl
     private static class Builder
             extends PlanVisitor<Context, JoinGraph>
     {
+        // TODO When com.facebook.presto.sql.planner.optimizations.EliminateCrossJoins is removed, remove 'shallow' flag
```

Comment:
Yes. Yes. Yes. But this approach does not fit into iterative approach either. Let's discuss it offline.